lib/elixir/lib/kernel.ex

```diff
@@ -2176,17 +2176,17 @@ defmodule Kernel do
       true ->
         raise ArgumentError, "cannot set attribute @#{name} inside function/macro"
       false ->
-        arg = case name do
-          :behavior ->
+        cond do
+          name == :behavior ->
             :elixir_errors.warn env.line, env.file,
                                 "@behavior attribute is not supported, please use @behaviour instead"
-          :doc       -> {env.line, arg}
-          :typedoc   -> {env.line, arg}
-          :moduledoc -> {env.line, arg}
-          _          -> arg
+          :lists.member(name, [:moduledoc, :typedoc, :doc]) ->
+            stack = env_stacktrace(env)
+            arg   = {env.line, arg}
+            quote do: Module.put_attribute(__MODULE__, unquote(name), unquote(arg), unquote(stack))
```

Comment:
@josevalim I've attempted to consolidate things here and pass only the stacktrace. However, passing the stack trace seems to cause an error when bootstrapping the `File` module:

```
Compiled lib/elixir/lib/path.ex
error: #{'__exception__' => true,
         '__struct__' => 'Elixir.CompileError',
         description => <<"invalid quoted expression: {File, :__MODULE__, 0, [file: \"lib/elixir/lib/file.ex\", line: 2]}">>,
         file => <<"/Users/dberkom/dev/elixir/lib/elixir/lib/file.ex">>,
         line => 0}
stacktrace: [{'Elixir.Kernel','@',1,[{file,"expanding macro"}]},
             {'Elixir.File','__MODULE__',0,
                            [{file,"/Users/dberkom/dev/elixir/lib/elixir/lib/file.ex"},
                             {line,2}]},
             {elixir,erl_eval,3,[{file,"src/elixir.erl"},{line,198}]},
             {elixir,eval_forms,4,[{file,"src/elixir.erl"},{line,186}]},
             {elixir_compiler,eval_compilation,3,
                              [{file,"src/elixir_compiler.erl"},{line,66}]},
             {elixir_compiler,quoted,3,
                              [{file,"src/elixir_compiler.erl"},{line,30}]},
             {elixir_compiler,core_file,1,
                              [{file,"src/elixir_compiler.erl"},{line,218}]},
             {elixir_compiler,'-core/0-lc$^0/1-0-',1,
                              [{file,"src/elixir_compiler.erl"},{line,214}]}]
make: *** [lib/elixir/ebin/Elixir.Kernel.beam] Error 1
```